Sources/MongoSwift/MongoCursor.swift

```diff
@@ -197,37 +197,58 @@ public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
      *   - `DecodingError` if an error occurs decoding the server's response.
      */
     public func nextOrError() throws -> T? {
-        if let next = self.next() {
-            return next
-        }
-        if let error = self.error {
-            throw error
+        let next = self.next()
+        if let next = next {
+            switch next {
+            case let .success(doc):
+                return doc
+            case let .failure(error):
+                throw error
+            }
         }
         return nil
     }
 
+    public func all() throws -> [T] {
+        return try self.map {
+            switch $0 {
+            case let .success(t):
+                return t
+            case let .failure(error):
+                throw error
+            }
+        }
+    }
+
     /// Returns the next `Document` in this cursor, or `nil`. After this function returns `nil`, the caller should use
     /// the `.error` property to check for errors. For tailable cursors, users should also check `isAlive` after this
     /// method returns `nil`, to determine if the cursor has the potential to return any more data in the future.
-    public func next() -> T? {
+    public func next() -> Result<T, Error>? {
         if case let .cached(doc) = self.cached {
             self.cached = .none
-            return doc
+            if let doc = doc {
+                return .success(doc)
+            }
+            return nil
         }
 
         // We already closed the mongoc cursor, either because we reached the end or encountered an error.
         guard case let .open(_, connection, client, session) = self.state else {
             self.error = ClosedCursorError
```

Comment:
I believe your changes here mean we can get rid of error caching altogether. We were doing it before because we didn't have a way to propagate any errors encountered when calling `next()`. Now we are returning those in the `Result`, so I think you can just remove the `error` property altogether and any places we are setting it.